### PR TITLE
STA-39: request timeout

### PR DESCRIPTION
**Pull request by dgellow**
Relevant issue: https://linear.app/stainless/issue/STA-39/timeouts

### CODE REVIEW

File: core.ts [hunk 1]
```diff
@@ -414,3 +443,13 @@ const safeJSON = (text: string) => {
 };
 
 const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
+
+const validatePositiveInteger = (name: string, n: number) => {
+  if (!Number.isInteger(n)) {
+    throw new Error(`${name} must be an integer`);
+  }
+  if (n < 0) {
+    throw new Error(`${name} must be a positive integer`);
```
Comment (dgellow) [inline core.ts:452]: I don't expect this error to be handled, I think using `Error` is ok.

File: core.ts [hunk 2]
```diff
@@ -60,8 +70,20 @@ export abstract class APIClient {
 
     this.debug('request', url, options, req.headers);
 
-    const response = await fetch(url, req).catch(() => null);
-
+    const fetchPromise = fetch(url, req).catch(() => null);
+    let timeoutId: NodeJS.Timeout | null;
```
Comment (dgellow) [inline core.ts:74]: I don't like the `NodeJS` dependency here.

File: README.md [hunk 3]
```diff
@@ -117,6 +117,23 @@ const lithic = new Lithic(process.env.LITHIC_API_KEY, {
 lithic.cards.list({page_size: 5}, {maxRetries: 5});
 ```
 
+### Timeouts
+
+By default requests timeout after 80 seconds. A different timeout value can set by using the client option `timeout`
+when instantiating the client, or in a request's options. The provided value is in millisecond, `timeout: 1000` will
+make request timeout after 1 second. E.g.:
```
Comment (dgellow) [inline README.md:124]:
```suggestion
make requests timeout after 1 second. E.g.:
```

**Review by rattrayalex — APPROVED** [review]

File: README.md [hunk 4]
```diff
@@ -105,7 +105,7 @@ Connection errors (for example, due to a network connectivity problem), 409 Conf
 and >=500 Internal errors will all be retried by default.
 
 You can disable or configure this behavior by passing `maxRetries: 0` to the client instantiation
-or in a request's options, e.g.;
+or in a request's options, e.g.:
```
Comment (rattrayalex) [inline README.md:108]:
```suggestion
or in a request's options, e.g.,
```
Nit, commas should follow the period with ie/eg

File: README.md [hunk 5]
```diff
@@ -117,6 +117,23 @@ const lithic = new Lithic(process.env.LITHIC_API_KEY, {
 lithic.cards.list({page_size: 5}, {maxRetries: 5});
 ```
 
+### Timeouts
+
+By default requests timeout after 80 seconds. A different timeout value can set by using the client option `timeout`
+when instantiating the client, or in a request's options. The provided value is in millisecond, `timeout: 1000` will
+make requests timeout after 1 second. E.g.:
```
Comment (rattrayalex) [inline README.md:124]:
```suggestion
make requests timeout after 1 second, e.g.,
```

Comment (rattrayalex) [reply README.md:124]: Actually, this sentence is redundant with teh code example and should be removed.

File: README.md [hunk 6]
```diff
@@ -117,6 +117,23 @@ const lithic = new Lithic(process.env.LITHIC_API_KEY, {
 lithic.cards.list({page_size: 5}, {maxRetries: 5});
 ```
 
+### Timeouts
+
+By default requests timeout after 80 seconds. A different timeout value can set by using the client option `timeout`
```
Comment (rattrayalex) [inline README.md:122]: I think this whole paragraph should just be:
```
Requests time out after 80 seconds by default; you may pass a `timeout` option when instantiating the client or sending a request to override this.
```

File: README.md [hunk 7]
```diff
@@ -117,6 +117,23 @@ const lithic = new Lithic(process.env.LITHIC_API_KEY, {
...
+```
+const lithic = new Lithic(process.env.LITHIC_API_KEY, {
+  timeout: 20 * 1000, // 20 seconds
+});
+
+// Or, configure per-request:
+lithic.cards.list({page_size: 5}, {timeout: 5 * 1000});
+```
+
+On timeout an error `APIConnectionTimeoutError` is thrown.
```
Comment (rattrayalex) [inline README.md:135]:
```suggestion
On timeout an `APIConnectionTimeoutError` is thrown.
```

Comment (rattrayalex) [reply README.md:135]: We may want to add: 

> Note that requests which time out will be [retried twice by default](#retries).

File: README.md (hunk 4 above)
Comment (dgellow) [reply README.md:108]: Oh, that's good to know 👍

File: README.md (hunk 6 above)
Comment (dgellow) [reply README.md:122]: Yep, that's good

**Pull request merged**